### PR TITLE
[PF-1576, PF-1596] Add Workspace.userFacingId and plumb through to API

### PR DESCRIPTION
**Pull request by melissachang**
This is a repeat of #625. I accidentally merged #625 into `mc/ufId` instead of `main`.

### CODE REVIEW

**Review by ddietterich — APPROVED** [review]

File: openapi/src/parts/workspace.yaml [hunk 1]
```diff
@@ -330,6 +330,11 @@ components:
           description: The ID of the workspace
           type: string
           format: uuid
+        userFacingId:
+          description: |
+            Human-settable, mutable id. Optional. If this isn't set, one will
+            be generated based on id.
```
Comment (jaycarlton) [inline openapi/src/parts/workspace.yaml:336]: generated based on the GUID?

File: openapi/src/parts/workspace.yaml [hunk 2]
```diff
@@ -416,12 +425,15 @@ components:
 
     WorkspaceDescription:
       type: object
-      required: [id]
+      required: [id, userFacingId]
       properties:
         id:
           description: The ID of the workspace
```
Comment (jaycarlton) [inline openapi/src/parts/workspace.yaml:431]: document that this one is immutable maybe.

File: service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java [hunk 3]
```diff
@@ -97,13 +100,29 @@ public UUID createWorkspace(Workspace workspace) {
...
-      throw new DuplicateWorkspaceException(
-          String.format(
-              "Workspace with id %s already exists - display name %s stage %s",
-              workspaceId,
-              workspace.getDisplayName().toString(),
-              workspace.getWorkspaceStage().toString()),
-          e);
+      // Workspace with workspace_id already exists.
+      if (e.getMessage()
+          .contains("duplicate key value violates unique constraint \"workspace_pkey\"")) {
```
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java:105]: This looks brittle; a change to the DB vendor or version might break the message. Not sure what your alternatives are though.

File: service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java [hunk 4]
```diff
@@ -193,14 +217,28 @@ public boolean updateWorkspace(
             "UPDATE workspace SET %s WHERE workspace_id = :workspace_id",
             DbUtils.setColumnsClause(params, "properties"));
 
-    int rowsAffected = jdbcTemplate.update(sql, params);
-    boolean updated = rowsAffected > 0;
+    int rowsAffected;
+    try {
+      rowsAffected = jdbcTemplate.update(sql, params);
+    } catch (DuplicateKeyException e) {
+      // Workspace with user_facing_id already exists.
```
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java:224]: This could be a duplicate of the guid as well.

File: service/src/main/resources/db/changesets/20220418_workspace_user_facing_id.yaml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    - addColumn:
+        tableName: workspace
+        columns:
+          - column:
+              name: user_facing_id
+              type: text
+              constraints:
+                unique: true
+              remarks: |
+                A user-settable, mutable id. See PF-1359. Required (a later
```
Comment (jaycarlton) [inline service/src/main/resources/db/changesets/20220418_workspace_user_facing_id.yaml:15]: We might not want the PR numbers in the public record here. I don't know where this string goes though.

File: service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java [hunk 6]
```diff
@@ -247,23 +264,43 @@ void testUpdateWorkspace() {
...
+    assertEquals(userFacingId, secondUpdatedWorkspace.getUserFacingId().orElse(null));
     assertEquals(name, secondUpdatedWorkspace.getDisplayName().orElse(null));
     assertEquals(otherDescription, secondUpdatedWorkspace.getDescription().orElse(null));
     assertEquals(propertyMap2, updatedWorkspace.getProperties());
 
     // Sending through empty strings and an empty map clears the values.
     Map<String, String> propertyMap3 = new HashMap<>();
     Workspace thirdUpdatedWorkspace =
-        workspaceService.updateWorkspace(USER_REQUEST, workspaceId, "", "", propertyMap3);
+        workspaceService.updateWorkspace(USER_REQUEST, workspaceId, "", "", "", propertyMap3);
```
Comment (jaycarlton) [inline service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java:278]: It would be more robust to test with separate strings, so we don't have a false negative if we cross wires somewhere.

File: service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java [hunk 7]
```diff
@@ -247,23 +264,43 @@ void testUpdateWorkspace() {
...
     assertEquals("", thirdUpdatedWorkspace.getDescription().orElse(null));
 
     assertThrows(
         MissingRequiredFieldException.class,
-        () -> workspaceService.updateWorkspace(USER_REQUEST, workspaceId, null, null, null));
+        () -> workspaceService.updateWorkspace(USER_REQUEST, workspaceId, null, null, null, null));
+  }
+
+  @Test
+  void testUpdateWorkspaceUserFacingIdAlreadyExistsRejected() {
```
Comment (jaycarlton) [inline service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java:289]: I thought you already tested this earlier.

**Review by jaycarlton — APPROVED** [review]
Got curious and left a few comments. Looks good!

File: openapi/src/parts/workspace.yaml (hunk 1 above)
Comment (melissachang) [reply openapi/src/parts/workspace.yaml]: Done

File: openapi/src/parts/workspace.yaml (hunk 2 above)
Comment (melissachang) [reply openapi/src/parts/workspace.yaml]: Done

File: service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java (hunk 4 above)
Comment (melissachang) [reply service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java:224]: This is `updateWorkspace()`, guid can't be updated. (There's no id field in UpdateWorkspaceRequestBody.)

File: service/src/main/resources/db/changesets/20220418_workspace_user_facing_id.yaml (hunk 5 above)
Comment (melissachang) [reply service/src/main/resources/db/changesets/20220418_workspace_user_facing_id.yaml:15]: There are a bunch of Jira ticket references in this repo...

File: service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java (hunk 6 above)
Comment (melissachang) [reply service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java]: Hmm, I made this test more realistic by passing `userFacingId` instead of `""`. (Controller validates that `userFacingId` can't be `""`. This test doesn't call controller, which is why `""` works.) So there's no more assertion for userFacingId.

File: service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java (hunk 7 above)
Comment (melissachang) [reply service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java:288]: `duplicateWorkspaceUserFacingIdRequestsRejected()` tests `createWorkspace()`, this tests `updateWorkspace()`. Both `createWorkspace()` and `updateWorkspace()` have validation logic.

File: service/src/main/resources/db/changesets/20220418_workspace_user_facing_id.yaml (hunk 5 above)
Comment (jaycarlton) [reply service/src/main/resources/db/changesets/20220418_workspace_user_facing_id.yaml:15]: Yeah, the other concern was the longevity of this text, but it's associated with the transaction only, so it's fine.
